### PR TITLE
Assign where compare meant false positives

### PR DESCRIPTION
**Pull request by MathiasVP**
Implements the techniques for reducing the false positive rate of the query `cpp/assign-where-compare-meant` as described in https://jira.semmle.com/browse/CPP-483.

Differences query: https://lgtm.com/query/8401788255056547968/

### CODE REVIEW

File: cpp/ql/test/query-tests/Likely Bugs/Likely Typos/AssignWhereCompareMeant/test.cpp [hunk 1]
```diff
@@ -98,3 +98,58 @@ void g(int *i_p, int cond) {
   if (y = 1) { // GOOD: y might not be initialized so it is probably intentional.
   }
 }
+
+template<typename>
+void h() {
+  int x;
+  if(x = 0) { // GOOD: x is not initialized so this is probably intensional
```
Comment (geoffw0) [inline cpp/ql/test/query-tests/Likely Bugs/Likely Typos/AssignWhereCompareMeant/test.cpp]:
```suggestion
  if(x = 0) { // GOOD: x is not initialized so this is probably intentional
```

File: cpp/ql/test/query-tests/Likely Bugs/Likely Typos/AssignWhereCompareMeant/test.cpp [hunk 2]
```diff
@@ -98,3 +98,58 @@ void g(int *i_p, int cond) {
...
+
+void f2() {
+  const char* sz = "abc";
+
+  if(sz = "def") { // GOOD: a == comparison with a string literal is probably not the intent here
+  }
+}
+
+void f3(int x, int y) {
+  if(x == 1 && (y = 2)) { // GOOD [FALSE POSITIVE]: the programmer seems to be okay with unparenthesized
```
Comment (geoffw0) [inline cpp/ql/test/query-tests/Likely Bugs/Likely Typos/AssignWhereCompareMeant/test.cpp]: I think this `[FALSE POSITIVE]` was fixed.

File: cpp/ql/test/query-tests/Likely Bugs/Likely Typos/AssignWhereCompareMeant/test.cpp [hunk 3]
```diff
@@ -98,3 +98,58 @@ void g(int *i_p, int cond) {
...
+  }
+
+  if((x == 1) && (y = 2)) { // BAD
+  }
+
+  long z = x;
+  if(((z == 42) || (y = 2)) && (x == 1)) { // BAD
+  }
+
+  if((y = 2) && (x == z || x == 1)) { // GOOD [FALSE POSITIVE]
```
Comment (geoffw0) [inline cpp/ql/test/query-tests/Likely Bugs/Likely Typos/AssignWhereCompareMeant/test.cpp]: I think this `[FALSE POSITIVE]` was fixed.

File: cpp/ql/test/query-tests/Likely Bugs/Likely Typos/AssignWhereCompareMeant/test.cpp [hunk 4]
```diff
@@ -98,3 +98,58 @@ void g(int *i_p, int cond) {
...
+  }
+
+  if(((x == 42) || x == 1) && (y = 2)) { // BAD
+  }
+
+  if(x == 10 || (x == 42 && x == 1) && (y = 2)) { // GOOD [FALSE POSITIVE]
```
Comment (geoffw0) [inline cpp/ql/test/query-tests/Likely Bugs/Likely Typos/AssignWhereCompareMeant/test.cpp]: I think this `[FALSE POSITIVE]` was fixed.

File: cpp/ql/src/Likely Bugs/Likely Typos/AssignWhereCompareMeant.ql [hunk 5]
```diff
@@ -38,14 +38,31 @@ abstract class BooleanControllingAssignment extends AssignExpr {
...
     this.getParent() instanceof BinaryLogicalOperation or
     exists(ConditionalExpr c | c.getCondition() = this)
   }
 
-  override predicate isWhitelisted() { this.getConversion().(ParenthesisExpr).isParenthesised() }
+  override predicate isWhitelisted() {
+    this.getConversion().(ParenthesisExpr).isParenthesised()
+    or
+    // whitelist this assignment if all comparison operations in the expression that this
+    // assignment is part of, are ot parenthesized. In that case it seems like programmer
```
Comment (geoffw0) [inline cpp/ql/src/Likely Bugs/Likely Typos/AssignWhereCompareMeant.ql]:
```suggestion
    // assignment is part of, are not parenthesized. In that case it seems like programmer
```

File: cpp/ql/test/query-tests/Likely Bugs/Likely Typos/AssignWhereCompareMeant/test.cpp (hunk 2 above)
Comment (MathiasVP) [reply cpp/ql/test/query-tests/Likely Bugs/Likely Typos/AssignWhereCompareMeant/test.cpp]: > I think this `[FALSE POSITIVE]` was fixed.

You're right!

**Pull request merged**